File: src/bot/commands/create.ts
import type { Context } from "grammy"
import { JiraAuthError, JiraNotFoundError } from "../../shared/errors"

// TODO: replace with import from './index' once section-07-registration-and-bot is complete
interface Clients {
  jira: {
    createIssue(title: string, description: string): Promise<{ key: string }>
  }
  claude: {
    ask(prompt: string): Promise<string>
  }
}

export const ENRICH_PROMPT_TEMPLATE = `You are a Jira ticket writer. Given the title and description provided by the user, write a well-formatted Jira ticket description in plain text.

<title>{title}</title>
<description>{description}</description>

Return only the formatted description. No preamble, no metadata.`

export const EXPAND_PROMPT_TEMPLATE = `You are a Jira ticket writer. Given the title provided by the user, write a Jira ticket description with 3-5 sentences and 3-5 acceptance criteria bullet points.

<title>{title}</title>

Return only the formatted description. No preamble, no metadata.`

export async function handleCreate(ctx: Context, clients: Clients): Promise<void> {
  const match = ((ctx.match as string) ?? "").trim()

  if (!match) {
    await ctx.reply("Usage: /create <title> [-- <description>]")
    return
  }

  try {
    await ctx.replyWithChatAction("typing")
    const typingInterval = setInterval(() => {
      ctx.replyWithChatAction("typing").catch(() => {})
    }, 4000)

    try {
      let title: string
      let description: string

      const separatorIdx = match.indexOf(" -- ")
      if (separatorIdx !== -1) {
        // Path A: enrich with Claude
        title = match.slice(0, separatorIdx).trim()
        const rawDescription = match.slice(separatorIdx + 4).trim()
        let enriched = rawDescription
        try {
          // Replace {description} before {title} to prevent cross-substitution injection
          const prompt = ENRICH_PROMPT_TEMPLATE.replace("{description}", rawDescription).replace(
            "{title}",
            title,
          )
          enriched = await clients.claude.ask(prompt)
        } catch {
          // silent fallback — use raw description
        }
        description = enriched
      } else {
        // Path B: expand with Claude
        title = match
        let expanded = ""
        try {
          const prompt = EXPAND_PROMPT_TEMPLATE.replace("{title}", title)
          expanded = await clients.claude.ask(prompt)
        } catch {
          // silent fallback — use empty description
        }
        description = expanded
      }

      const issue = await clients.jira.createIssue(title, description)
      await ctx.reply(`Created: ${issue.key}`)
    } finally {
      clearInterval(typingInterval)
    }
  } catch (err) {
    if (err instanceof JiraAuthError) {
      await ctx.reply("Authentication failed. Please check your Jira API token.")
      return
    }
    if (err instanceof JiraNotFoundError) {
      await ctx.reply("Jira resource not found.")
      return
    }
    const message = err instanceof Error ? err.message : String(err)
    console.log({ event: "error", command: "create", errorMessage: message })
    await ctx.reply("Something went wrong. Please try again.")
  }
}

File: src/bot/commands/create.rs
use std::sync::Arc;

use anyhow::Result;
use serde_json::json;

use crate::bot::state::JiraPendingAction;
use crate::bot::AppState;
use crate::channel::{Button, ChannelSender};
use crate::claude::types::AskOptions;
use crate::config::loader::load_config;
use crate::shared::paths::PATHS;

/// Description-drafting instructions used when the project has no
/// `project_ticket_templates` entry configured, or its file can't be read.
const DEFAULT_DESCRIPTION_TEMPLATE: &str = "\
Write a professional description including: brief overview, acceptance criteria as a bullet list, \
and relevant technical notes. Generate it from the title.";

/// Reads the Markdown template file configured for `project_key`, if any.
/// Relative paths resolve against the devm8 config directory.
fn load_description_template(project_key: &str) -> Option<String> {
    let path = load_config(None)
        .ok()?
        .project_ticket_templates
        .get(project_key)?
        .clone();
    let path = std::path::Path::new(&path);
    let resolved = if path.is_absolute() {
        path.to_path_buf()
    } else {
        PATHS.config_dir.join(path)
    };
    std::fs::read_to_string(&resolved).ok().map(|s| s.trim().to_string())
}

fn build_improve_prompt(project_key: &str, title: &str) -> String {
    let template =
        load_description_template(project_key).unwrap_or_else(|| DEFAULT_DESCRIPTION_TEMPLATE.to_string());

    format!(
        "You are a technical project manager improving a Jira ticket.\n\n\
         Project: {project_key}\n\
         Original title: {title}\n\n\
         Your tasks:\n\
         1. Correct and improve the title — fix grammar, spelling, and clarity; keep it concise (under 80 chars).\n\
         2. {template}\n\n\
         Respond in this exact format with nothing else before or after:\n\
         TITLE: <corrected title>\n\n\
         DESCRIPTION:\n\
         <description here>"
    )
}

fn parse_claude_response(response: &str) -> (String, String) {
    let title = response
        .lines()
        .find(|l| l.starts_with("TITLE:"))
        .map(|l| l["TITLE:".len()..].trim().to_string())
        .unwrap_or_default();

    let desc = response
        .find("DESCRIPTION:\n")
        .map(|idx| response[idx + "DESCRIPTION:\n".len()..].trim().to_string())
        .unwrap_or_default();

    (title, desc)
}

/// Step 1: receives the raw title, corrects it with Claude, generates a description
/// suggestion, then shows it to the user with a "Use this" button.
pub async fn handle_create_suggest(
    sender: Arc<dyn ChannelSender>,
    chat_id: &str,
    state: Arc<AppState>,
    _user_id: &str,
    project_key: String,
    title: String,
) -> Result<()> {
    if title.is_empty() {
        sender
            .send(
                chat_id,
                "Send the issue title:\n<code>New login page</code>",
            )
            .await?;
        return Ok(());
    }

    let thinking_ref = sender
        .send(chat_id, "Correcting title and generating description...")
        .await?;
    let _typing = sender.start_typing(chat_id);

    let prompt = build_improve_prompt(&project_key, &title);

    let claude_output = match state.ai.ask(&prompt, AskOptions::default()).await {
        Ok((text, _)) => text,
        Err(e) => {
            state.logger.error(
                &format!("create: Claude error: {e}"),
                Some(&json!({ "title": &title })),
            );
            sender
                .edit_text(&thinking_ref, &format!("Claude error: {e}"))
                .await?;
            return Ok(());
        }
    };

    let (corrected_title, suggested_desc) = parse_claude_response(&claude_output);
    let corrected_title = if corrected_title.is_empty() {
        title
    } else {
        corrected_title
    };

    state
        .chat_states
        .entry(chat_id.to_string())
        .or_default()
        .pending_jira_action = Some(JiraPendingAction::CreateDescription(
        project_key,
        corrected_title.clone(),
        suggested_desc.clone(),
    ));

    let keyboard = vec![vec![Button::new(
        "\u{2705} Use this description",
        "jira:create_confirm",
    )]];

    sender
        .edit_with_keyboard(
            &thinking_ref,
            &format!(
                "Corrected title: <b>{}</b>\n\nSuggested description:\n<pre>{}</pre>\n\n\
                 Tap <b>Use this description</b> or send your own below:",
                sender.escape(&corrected_title),
                sender.escape(&suggested_desc)
            ),
            keyboard,
        )
        .await?;

    Ok(())
}

/// Step 2: creates the Jira issue with the final (confirmed or custom) description.
pub async fn handle_create_confirm(
    sender: Arc<dyn ChannelSender>,
    chat_id: &str,
    state: Arc<AppState>,
    user_id: &str,
    project_key: &str,
    title: &str,
    description: &str,
) -> Result<()> {
    state.logger.info(
        "create: creating Jira issue",
        Some(&json!({ "project": project_key, "title": title })),
    );

    let thinking_ref = sender.send(chat_id, "Creating issue...").await?;

    let Some(jira) = state.jira_for_user(user_id).await else {
        sender
            .edit_text(
                &thinking_ref,
                "Please set up your Jira account first. Use /jira \u{2192} My Jira.",
            )
            .await?;
        return Ok(());
    };
    let issue = match jira.create_issue(project_key, title, description).await {
        Ok(issue) => issue,
        Err(e) => {
            state.logger.error(
                &format!("create: Jira error: {e}"),
                Some(&json!({ "project": project_key, "title": title })),
            );
            sender
                .edit_text(&thinking_ref, &format!("Jira error: {e}"))
                .await?;
            return Ok(());
        }
    };

    state.logger.info(
        "create: issue created",
        Some(&json!({ "key": &issue.key, "url": &issue.url })),
    );

    sender
        .edit_text(
            &thinking_ref,
            &format!(
                "Created: <a href=\"{}\">{}</a> \u{2014} {}",
                issue.url, issue.key, issue.summary
            ),
        )
        .await?;

    Ok(())
}
